network/peer.go

```diff
@@ -836,3 +879,45 @@ func (p *peer) getIP() utils.IPDesc {
 	defer p.ipLock.RUnlock()
 	return p.ip
 }
+
+// addAlias marks that we have found another
+// IP that we can connect to this peer at.
+//
+// assumes stateLock is held
+func (p *peer) addAlias(ip utils.IPDesc) {
+	p.ipLock.Lock()
+	defer p.ipLock.Unlock()
+
+	p.net.peerAliasIPs[ip.String()] = struct{}{}
+	p.aliases = append(p.aliases, alias{
+		ip:    ip,
+		added: p.net.clock.Time(),
+	})
+}
+
+// releaseAliases removes enables network to
+// reconnect on aliases that have timed out. To
+// force all aliases to be removed, the caller
+// should set now to be the zero time.
+func (p *peer) releaseAliases(now time.Time, holdsStateLock bool) {
+	p.ipLock.Lock()
+	defer p.ipLock.Unlock()
+
+	for len(p.aliases) > 0 {
+		next := p.aliases[0]
+		if !now.IsZero() && now.Sub(next.added) < p.net.peerAliasTimeout {
+			return
+		}
+		p.aliases = p.aliases[1:]
+
+		if !holdsStateLock {
+			p.net.stateLock.Lock()
+		}
+		delete(p.net.peerAliasIPs, next.ip.String())
```

Comment:
The lock needs to be held in network code to safely do this delete.

Reply:
Yep! The network lock is acquired prior to this line if called by `monitorAliases` (https://github.com/ava-labs/avalanchego/pull/746/files#diff-b06a14805cb9806d82ef18d987b32228ab9b36a4e679c7c128efd711038991b2R157)  or is already held when this function is called by `disconnected` (https://github.com/ava-labs/avalanchego/pull/746/files#diff-96a7f5c352cd3d2d1ee3c7c55ed95e668c6edb28a4fa337e6c60270471bdab61R1289).